ephios/extra/auth.py

```diff
@@ -42,7 +44,24 @@ def update_user(self, user, claims):
             except ValueError:
                 pass
         user.save()
-        if hasattr(self, "provider") and self.provider.default_groups.exists():
+        if self.provider.group_claim:
+            groups = []
+            groups_in_claims = (
+                reduce(
+                    lambda d, key: d.get(key, None) if isinstance(d, dict) else None,
+                    self.provider.group_claim.split("."),
+                    claims,
+                )
+                or []
+            )
+            for group_name in groups_in_claims:
+                try:
+                    groups.append(Group.objects.get(name__iexact=group_name))
+                except Group.DoesNotExist:
+                    if self.provider.create_missing_groups:
+                        groups.append(Group.objects.create(name=group_name))
+            user.groups.set(groups)
+        elif self.provider.default_groups.exists():
```

Comment:
It might be unexpected from an administrator's perspective that the default groups will not be assigned when using the groups from the claim. I think you should either change this line from `elif` to an independent `if` (which should be fine, since you're using `add`for the default groups) or state in the documentation and field help texts that `default_groups` and `group_claim` are exclusive.

Reply:
I stated that in the documentation (see https://ephios--1115.org.readthedocs.build/en/1115/admin/additional_features/oidc.html from the readthedocs status check). For me it seemed more intuitive to have the IDP in total control of the groups if it supports groups. I'm not sure what the usecase for adding a group to _everyone_ from an IDP would be. @felixrindt thoughts?

Reply:
> I stated that in the documentation (see https://ephios--1115.org.readthedocs.build/en/1115/admin/additional_features/oidc.html from the readthedocs status check).

Are you referring to _The user will be a member of all groups that are listed in the claim, meaning they will be removed from groups that they have been manually assigned to in ephios and that are not listed in the claim_?

From my perspective, the default groups are rather _automatically_ assigned to all users than _manually_ assigned, that's why I thought the docs aren't mentioning it.

> I'm not sure what the usecase for adding a group to _everyone_ from an IDP would be.

When having multiple IDPs, it might be handy to automatically assign a group to each user, identifying the IDP this account is taken from. The IDP might not assign such a role, as from the IDP's perspective, every user would have this role, making it redundant.

Reply:
> I'm not sure what the usecase for adding a group to _everyone_ from an IDP would be. @felixrindt thoughts?

As stated in my other comment, I think your implementation is not flexible enough. Take the HPI IDP for example. FSR uses it but needs additional groups in ephios (e.g. FSR members). Also, an IDP might provide no group for some user, but we want to have them in at least one group, so default groups could still be useful.

Reply:
I think your example is not a good one, because it does not support exposing groups via claims. The feature is suited for a scenario which the FSR will most likely have soon: a Keycloak controlled by the FSR where they can set arbitrary groups, so especially one for FSR members. I think it would be really confusing to assign some groups in ephios and some in the IDP, as there is no reason to not just assign everything in the IDP. I think the usecase provided by @lukasrad02 is more suited. If I use two different IDPs, I could use the default groups to distinguish between them. (e.g for the current staging setup have one group for everyone signing in from the HPI). However, this still seems like some niche use case for me.

Reply:
we decided to add default groups also if oidc claims are used.